src/build.cc

```diff
@@ -768,32 +774,40 @@ bool Builder::FinishCommand(CommandRunner::Result* result, string* err) {
     }
 
     if (node_cleaned) {
-      TimeStamp restat_mtime = 0;
-      // If any output was cleaned, find the most recent mtime of any
-      // (existing) non-order-only input or the depfile.
-      for (vector<Node*>::iterator i = edge->inputs_.begin();
-           i != edge->inputs_.end() - edge->order_only_deps_; ++i) {
-        TimeStamp input_mtime = disk_interface_->Stat((*i)->path(), err);
-        if (input_mtime == -1)
-          return false;
-        if (input_mtime > restat_mtime)
-          restat_mtime = input_mtime;
-      }
-
-      string depfile = edge->GetUnescapedDepfile();
-      if (restat_mtime != 0 && deps_type.empty() && !depfile.empty()) {
-        TimeStamp depfile_mtime = disk_interface_->Stat(depfile, err);
-        if (depfile_mtime == -1)
-          return false;
-        if (depfile_mtime > restat_mtime)
-          restat_mtime = depfile_mtime;
-      }
-
       // The total number of edges in the plan may have changed as a result
       // of a restat.
       status_->PlanHasTotalEdges(plan_.command_edge_count());
+    }
 
-      output_mtime = restat_mtime;
+    // Use the time from the most recent input that was computed when the edge was
+    // started, not the mtime of the edge's latest output as it is now. There could have
+    // been other edges that restat'd the input node and detected a change, but for *this*
+    // edge, we want the mtime as it was when the command began.
+    //
+    // Generator edges should just use the edge's latest output mtime since it's possible that
+    // the generator rule could create/update implicit inputs as part of the command, and
+    // we don't want to get into an infinite loop.
+    if (!edge->GetBindingBool("generator")) {
+      record_mtime = edge->most_recent_input_mtime_;
+
+      // If there were any added deps, compute the most recent input mtime
+      for (vector<Node*>::iterator i = deps_nodes.begin();
+          i != deps_nodes.end(); ++i) {
+        (*i)->StatIfNecessary(disk_interface_, err);
```

Comment:
What if the newly-known input was modified externally between the time that the edge's command actually used it, and the time that we are now stat-ing it?

Reply:
Since we didn't know about it when we started, the best we can do is get the time now. It's why the PR isn't 100% robust around external input modification during the build. The only way to get 100% correctness is to use dyndep to discover all inputs before the actual edge command is started. To be fair, doing it this way is no worse than what currently exists in ninja (other than the possibility of an unneeded rebuild on a subsequent run - if this input wasn't known to this edge specifically but was known during the initial build plan generation).

We can't make this `Stat()` as it defeats the purpose of the PR. Since `deps_nodes` contains all dependencies from the "special" dep file (not only those that weren't known at edge command start time), using `Stat()` here would mean that changes to the input while the edge command is run would be lost on future builds.

A better way to do this would be to somehow track which deps were discovered that are "new" and were not counted for during the initial edge start input dep list. We could `Stat()` only those, and simply bypass doing anything with the ones we know we counted for when edge was started.

Unfortunately, the `RecordDeps()` function used below this short circuits actually figuring out _which_ deps are new in favor of simply replacing the list with the new one if they are detected as different. I could take a crack at improving this dependency discovery/updating to be a bit more intelligent without sacrificing any performance, but I'm not sure where that will take me.